Convert Mixer self-monitoring to use OpenCensus libraries

douglas-reid opened the pull request:
This PR is part of an exploration of moving Istio self-monitoring from pure Prometheus libraries to OpenCensus (OC) libraries. The ultimate goal being explored is the possibility of creating an Istio OpenCensus exporter mechanism that would allow flexible export of Istio self-monitoring metrics to different backends (cloudwatch, datadog, stackdriver, etc.).

This PR does the following: 
 - adds an OC gRPC stats handler to replace the Prometheus gRPC interceptor
 - tries to track down and remove all self-monitoring bits using Prometheus libraries (`checkcache`, `runtime`)
 - configures a Prometheus exporter for now (on same self-monitoring port)

This PR does *not* replace the Prometheus collectors that provide additional metrics (not from Mixer). These include:
 - Golang collector (for gauges on go-routines, etc.)
 - Process collector (linux process monitoring bits)

Reviewers:
I am interested in feedback on a few items. First, does this library change make sense? Second, is the way that OC is being used seem appropriate?

Important items to note when reviewing:
OC models things differently to prometheus and this results in some interesting changes:
 - For gauge aggregations, there isn't any way currently to express "last value +/- delta". So, in places where the codebase was using that sort of logic, new logic was added to separately track the value.
 - In establishing views, I've chosen here to mostly use `view.Count`. This is primarily because that results in exporting Prometheus metrics of type `Counter`. Using `view.Sum` results in untyped prometheus metrics. This ultimately impacts how we use the Stats, as `view.Count` records the number of measurements, not the total of those measurements. I *think* this is the right thing to do for now, but maybe not. The Prometheus docs include the following related documentation:
> The Prometheus server does not yet make use of the type information and flattens all data into untyped time series. This may change in the future.
- Because OC does not export stats for which no measurements that have been recorded, the overall output to prometheus will be different than before (we won't see `0`s for things that haven't happened, etc.)

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **douglas-reid**

If they are not already assigned, you can assign the PR to them by writing `/assign @douglas-reid` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/istio/istio/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["douglas-reid"]} -->

douglas-reid commented:
cc: @ramonza

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-dashboard.sh | ffbed4cca22141c0a177ddbe83bd0ecc3fe59b56 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-dashboard/2406/) | `/test e2e-dashboard`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

geeknoid commented on mixer/pkg/checkcache/cache.go:
```diff
@@ -81,65 +108,20 @@ func New(capacity int32) *Cache {
 
 // Get looks up an attribute bag in the cache.
 func (cc *Cache) Get(attrs attribute.Bag) (Value, bool) {
+	defer func() { cc.recordStats() }()
```
Here and elsewhere in this file, I'm not thrilled with "defer" here since it induces an allocation (last I looked)

douglas-reid replied:
Removed.

geeknoid commented on mixer/pkg/server/server.go:
```diff
@@ -218,6 +212,20 @@ func newServer(a *Args, p *patchTable) (*Server, error) {
+
+	exporter, err := prometheus.NewExporter(prometheus.Options{Registry: oprometheus.DefaultRegisterer.(*oprometheus.Registry)})
+	if err != nil {
+		return nil, fmt.Errorf("could not build prometheus exporter: %v", err)
```
Need to add a unit test to cover this new code.

geeknoid commented on mixer/pkg/server/server.go:
```diff
@@ -218,6 +212,20 @@ func newServer(a *Args, p *patchTable) (*Server, error) {
+
+	// Register the views to collect server request count.
+	if err := view.Register(ocgrpc.DefaultServerViews...); err != nil {
+		return nil, fmt.Errorf("could not register default server views: %v", err)
```
Need unit test for this.

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-dashboard.sh | ffbed4cca22141c0a177ddbe83bd0ecc3fe59b56 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-dashboard/2406/) | `/test e2e-dashboard`
prow/istio-unit-tests.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-unit-tests/11961/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-dashboard.sh | ffbed4cca22141c0a177ddbe83bd0ecc3fe59b56 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-dashboard/2406/) | `/test e2e-dashboard`
prow/istio-unit-tests.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-unit-tests/11961/) | `/test istio-unit-tests`
prow/e2e-bookInfoTests-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-bookInfoTests-envoyv2-v1alpha3/5081/) | `/test e2e-bookInfo-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-dashboard.sh | ffbed4cca22141c0a177ddbe83bd0ecc3fe59b56 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-dashboard/2406/) | `/test e2e-dashboard`
prow/istio-unit-tests.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-unit-tests/11961/) | `/test istio-unit-tests`
prow/e2e-bookInfoTests-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-bookInfoTests-envoyv2-v1alpha3/5081/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/e2e-mixer-no_auth.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-mixer-no_auth/2870/) | `/test e2e-mixer-no_auth`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-unit-tests/11961/) | `/test istio-unit-tests`
prow/e2e-bookInfoTests-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-bookInfoTests-envoyv2-v1alpha3/5081/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/e2e-mixer-no_auth.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-mixer-no_auth/2870/) | `/test e2e-mixer-no_auth`
prow/e2e-dashboard.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-dashboard/2718/) | `/test e2e-dashboard`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-unit-tests/11961/) | `/test istio-unit-tests`
prow/e2e-bookInfoTests-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-bookInfoTests-envoyv2-v1alpha3/5081/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/e2e-mixer-no_auth.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-mixer-no_auth/2870/) | `/test e2e-mixer-no_auth`
prow/e2e-dashboard.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-dashboard/2718/) | `/test e2e-dashboard`
prow/e2e-simpleTests.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-simpleTests/8084/) | `/test e2e-simple`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-unit-tests/11961/) | `/test istio-unit-tests`
prow/e2e-bookInfoTests-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-bookInfoTests-envoyv2-v1alpha3/5081/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/e2e-mixer-no_auth.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-mixer-no_auth/2870/) | `/test e2e-mixer-no_auth`
prow/e2e-dashboard.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-dashboard/2718/) | `/test e2e-dashboard`
prow/e2e-simpleTests.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-simpleTests/8084/) | `/test e2e-simple`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-pilot-e2e-envoyv2-v1alpha3/5880/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-unit-tests/11961/) | `/test istio-unit-tests`
prow/e2e-bookInfoTests-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-bookInfoTests-envoyv2-v1alpha3/5081/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/e2e-mixer-no_auth.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-mixer-no_auth/2870/) | `/test e2e-mixer-no_auth`
prow/e2e-dashboard.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-dashboard/2718/) | `/test e2e-dashboard`
prow/e2e-simpleTests.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-simpleTests/8084/) | `/test e2e-simple`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-pilot-e2e-envoyv2-v1alpha3/5880/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-multicluster-e2e.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-pilot-multicluster-e2e/789/) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-bookInfoTests-envoyv2-v1alpha3/5081/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/e2e-mixer-no_auth.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-mixer-no_auth/2870/) | `/test e2e-mixer-no_auth`
prow/e2e-dashboard.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-dashboard/2718/) | `/test e2e-dashboard`
prow/e2e-simpleTests.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-simpleTests/8084/) | `/test e2e-simple`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-pilot-e2e-envoyv2-v1alpha3/5880/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-multicluster-e2e.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-pilot-multicluster-e2e/789/) | `/test istio-pilot-multicluster-e2e`
prow/istio-unit-tests.sh | b3d425f914eaeb7a47a08aaaeec0e2060132830e | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-unit-tests/11987/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

douglas-reid commented on mixer/pkg/checkcache/cache.go:
```diff
@@ -72,6 +76,29 @@ type Value struct {
+	missesTotal    = stats.Int64("mixer/checkcache/cache_misses_total", "The number of times a cache lookup operation failed to find an entry in the cache.", stats.UnitDimensionless)
+	evictionsTotal = stats.Int64("mixer/checkcache/cache_evictions_total", "The number of entries that have been evicted from the cache.", stats.UnitDimensionless)
+
+	writesView    = newView(writesTotal, []tag.Key{}, view.Sum())
```
Great catch!

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-bookInfoTests-envoyv2-v1alpha3/5081/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/e2e-mixer-no_auth.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-mixer-no_auth/2870/) | `/test e2e-mixer-no_auth`
prow/e2e-dashboard.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-dashboard/2718/) | `/test e2e-dashboard`
prow/e2e-simpleTests.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-simpleTests/8084/) | `/test e2e-simple`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-pilot-e2e-envoyv2-v1alpha3/5880/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-multicluster-e2e.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-pilot-multicluster-e2e/789/) | `/test istio-pilot-multicluster-e2e`
prow/istio-unit-tests.sh | 8dac330377163ec8953e5ea427a98fc4e416d081 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-unit-tests/11991/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-bookInfoTests-envoyv2-v1alpha3/5081/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/e2e-mixer-no_auth.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-mixer-no_auth/2870/) | `/test e2e-mixer-no_auth`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-pilot-e2e-envoyv2-v1alpha3/5880/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-multicluster-e2e.sh | 950fb11bb7985164c73b3630ef994486f861794a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-pilot-multicluster-e2e/789/) | `/test istio-pilot-multicluster-e2e`
prow/istio-unit-tests.sh | 8dac330377163ec8953e5ea427a98fc4e416d081 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-unit-tests/11991/) | `/test istio-unit-tests`
prow/e2e-dashboard.sh | 8dac330377163ec8953e5ea427a98fc4e416d081 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-dashboard/2745/) | `/test e2e-dashboard`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-dashboard.sh | 8dac330377163ec8953e5ea427a98fc4e416d081 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-dashboard/2745/) | `/test e2e-dashboard`
prow/istio-pilot-multicluster-e2e.sh | 73b02cac8167f4e764bd5922f660de9d30191a83 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-pilot-multicluster-e2e/815/) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | 73b02cac8167f4e764bd5922f660de9d30191a83 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-pilot-multicluster-e2e/815/) | `/test istio-pilot-multicluster-e2e`
prow/e2e-dashboard.sh | 73b02cac8167f4e764bd5922f660de9d30191a83 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-dashboard/2755/) | `/test e2e-dashboard`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

douglas-reid commented:
Reviewers: PTAL.

I believe I have addressed the existing comments (adding unit tests, etc.). I have also updated the `e2e_dashboard` test (passes on local clusters) and the Mixer dashboard that is exercised by that test.

douglas-reid commented:
@geeknoid was there something you had in mind re: metrics used by our deps?  I was not aware of any dependencies that were using prometheus to monitor anything prior to this PR.

I don't know that we _need_ to solve that particular issue right now, but it does raise an interesting question of package instrumentation and playing well with others. @ramonza does OC have any thoughts on such issues?

douglas-reid commented:
/test istio-unit-tests

and, of course, linting produces different results on every run... sigh.

geeknoid commented:
Sorry for my unclear comment. I was referring to this from your PR comment:

> This PR does not replace the Prometheus collectors that provide additional metrics (not from Mixer). These include:
> 
> Golang collector (for gauges on go-routines, etc.)
> Process collector (linux process monitoring bits)
> 

Can these be addressed too at some point in a systematic way?

On a related note, the ControlZ library is currently mining Prometheus metrics for display in the UI. What happens in ControlZ after this PR goes in?

douglas-reid commented:
@geeknoid ah, gotcha. Yes, probably. But rather than duplicate that functionality in this PR (the key is really finding some hook to generate them on the fly the same way prometheus does, I believe), I've skipped over providing those in an OpenCensus way.  I think, ultimately, it probably makes more sense for OpenCensus to provide those measures than for Istio to build them. For now, skipping them seems fine. @ramonza thoughts here?

RE ControlZ:

It still works. We use the default registry (gatherer) for the prometheus exporter here, and that is what is currently powering the `metrics` portion of ControlZ.

geeknoid commented:
/approve
/lgtm


On Wed, Aug 22, 2018 at 8:42 AM Douglas Reid <notifications@github.com>
wrote:

> @geeknoid <https://github.com/geeknoid> ah, gotcha. Yes, probably. But
> rather than duplicate that functionality in this PR (the key is really
> finding some hook to generate them on the fly the same way prometheus does,
> I believe), I've skipped over providing those in an OpenCensus way. I
> think, ultimately, it probably makes more sense for OpenCensus to provide
> those measures than for Istio to build them. For now, skipping them seems
> fine. @ramonza <https://github.com/ramonza> thoughts here?
>
> RE ControlZ:
>
> It still works. We use the default registry (gatherer) for the prometheus
> exporter here, and that is what is currently powering the metrics portion
> of ControlZ.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/7989#issuecomment-415078190>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AVucHYZ8x4mX7qP9bQWH-hf9VylOus0yks5uTXvvgaJpZM4WAS-9>
> .
>

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/7989#issuecomment-415097685" title="LGTM">geeknoid</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/istio/istio/blob/master/OWNERS)~~ [geeknoid]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

istio-testing commented:
New changes are detected. LGTM label has been removed.

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-mixer-no_auth.sh | 569c02d21564e5babf605254e13610447f8ba2dc | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/e2e-mixer-no_auth/2975/) | `/test e2e-mixer-no_auth`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

douglas-reid commented:
Thanks for all the reviews!  I've fixed up all of the linter issues and dep diff stuff now. PTAL for final approval.

geeknoid commented:
/lgtm
/approve

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/7989#issuecomment-415462028" title="Approved">geeknoid</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/istio/istio/blob/master/OWNERS)~~ [geeknoid]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | c76dd928d563493044bbecee541364ced3c81d89 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-pilot-multicluster-e2e/933/) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | c76dd928d563493044bbecee541364ced3c81d89 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7989/istio-pilot-multicluster-e2e/938/) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->
